File: sdks/typescript/src/resources/gift_cards.ts
import { AxiosInstance } from 'axios';
import { PaginatedResponse } from '../types';

export interface GiftCard {
    id: string;
    code: string;
    balance: number;
    initial_value: number;
    status: string;
    created_at: string;
    updated_at: string;
}

export class GiftCards {
    constructor(private client: AxiosInstance) { }

    async list(page: number = 1, pageSize: number = 20, status?: string): Promise<PaginatedResponse<GiftCard>> {
        const params: any = { page, page_size: pageSize };
        if (status) params.status = status;
        const response = await this.client.get('/gift-cards/', { params });
        return response.data;
    }

    async get(cardId: string): Promise<GiftCard> {
        const response = await this.client.get(`/gift-cards/${cardId}/`);
        return response.data;
    }

    async create(data: Partial<GiftCard>): Promise<GiftCard> {
        const response = await this.client.post('/gift-cards/', data);
        return response.data;
    }

    async redeem(cardId: string, amount: number): Promise<GiftCard> {
        const response = await this.client.post(`/gift-cards/${cardId}/redeem/`, { amount });
        return response.data;
    }

    async checkBalance(cardCode: string): Promise<GiftCard> {
        const response = await this.client.get('/gift-cards/check-balance/', {
            params: { code: cardCode }
        });
        return response.data;
    }
}

File: sdks/rust/src/resources/gift_cards.rs
use crate::models::{PaginatedResponse, GiftCard};
use crate::Puxbay;
use crate::PuxbayError;
use reqwest::Method;
use serde::Serialize;
use std::collections::HashMap;

pub struct GiftCardsResource<'a> {
    pub(crate) client: &'a Puxbay,
}

impl<'a> GiftCardsResource<'a> {
    pub fn new(client: &'a Puxbay) -> Self {
        Self { client }
    }

    pub async fn list(&self, page: i32, status: Option<&str>) -> Result<PaginatedResponse<GiftCard>, PuxbayError> {
        let mut query = format!("gift-cards/?page={}", page);
        if let Some(s) = status {
            query.push_str(&format!("&status={}", s));
        }
        
        self.client.request(
            Method::GET,
            &query,
            None::<()>,
        ).await
    }

    pub async fn get(&self, card_id: &str) -> Result<GiftCard, PuxbayError> {
        self.client.request(
            Method::GET,
            &format!("gift-cards/{}/", card_id),
            None::<()>,
        ).await
    }

    pub async fn create(&self, card: &GiftCard) -> Result<GiftCard, PuxbayError> {
        self.client.request(
            Method::POST,
            "gift-cards/",
            Some(card),
        ).await
    }

    pub async fn redeem(&self, card_id: &str, amount: f64) -> Result<GiftCard, PuxbayError> {
        #[derive(Serialize)]
        struct RedeemRequest {
            amount: f64,
        }

        self.client.request(
            Method::POST,
            &format!("gift-cards/{}/redeem/", card_id),
            Some(&RedeemRequest { amount }),
        ).await
    }

    pub async fn check_balance(&self, code: &str) -> Result<HashMap<String, f64>, PuxbayError> {
        self.client.request(
            Method::GET,
            &format!("gift-cards/check-balance/?code={}", code),
            None::<()>,
        ).await
    }
}
